catalyst-toolbox-cli/src/cli/rewards/voters.rs

```diff
@@ -0,0 +1,123 @@
+use super::Error;
+use jcli_lib::jcli_lib::block::Common;
+
+use structopt::StructOpt;
+
+use chain_addr::{Discrimination, Kind};
+use chain_impl_mockchain::vote::CommitteeId;
+use jormungandr_lib::interfaces::{Address, Block0Configuration, Initial};
+use std::collections::{HashMap, HashSet};
+use std::ops::{Div, Mul};
+
+const ADA_TO_LOVELACE_FACTOR: u64 = 1_000_000;
+
+#[derive(StructOpt)]
+#[structopt(rename_all = "kebab-case")]
+pub struct VotersRewards {
+    #[structopt(flatten)]
+    common: Common,
+    /// Reward (in LOVELACE) to be distributed
+    #[structopt(long = "total-rewards")]
+    total_rewards: u64,
+}
+
+fn calculate_stake<'address>(
+    committee_keys: &HashSet<Address>,
+    block0: &'address Block0Configuration,
+) -> (u64, HashMap<&'address Address, u64>) {
+    let mut total_stake: u64 = 0;
+    let mut stake_per_voter: HashMap<&'address Address, u64> = HashMap::new();
+
+    for fund in &block0.initial {
+        match fund {
+            Initial::Fund(fund) => {
+                for utxo in fund {
+                    if !committee_keys.contains(&utxo.address) {
+                        let value: u64 = utxo.value.into();
+                        total_stake += value;
+                        let entry = stake_per_voter.entry(&utxo.address).or_default();
+                        *entry += value;
+                    }
+                }
+            }
+            Initial::Cert(_) => {}
+            Initial::LegacyFund(_) => {}
+        }
+    }
+    (total_stake, stake_per_voter)
+}
+
+/// Rewards are u64 for keeping the it we would calculate the inverse total_stake/voter_stake
+fn calculate_inverse_reward_share<'address>(
+    total_stake: u64,
+    stake_per_voter: &HashMap<&'address Address, u64>,
+) -> HashMap<&'address Address, u64> {
+    stake_per_voter
+        .iter()
+        .map(|(k, v)| (*k, total_stake.div(v)))
+        .collect()
+}
+
+/// caculate total reward from the inverse of the the reward share
+fn calculate_total_reward(share: u64, total_reward: u64) -> f64 {
+    (total_reward as f64).div(share as f64)
+}
+
+fn write_rewards_results(
+    common: Common,
+    stake_per_voter: &HashMap<&Address, u64>,
+    results: &HashMap<&Address, u64>,
+    total_rewards: u64,
+) -> Result<(), Error> {
+    let writer = common.open_output()?;
+    let header = [
+        "Address",
+        "Stake of the voter (ADA)",
```

Comment:
Also, I think this is in lovelaces. Unless you are really sure this is in ADA, but I doubt the snapshot is in adas.

Reply:
I had a mess with this. It is in ADA or Lovelaces? I thought it was in lovelaces first then in ADA and I have been switching between them.

Reply:
Well, I didn't actually check the snapshot code this time, but I always thought it is in lovelaces... It'd be weird if we changed it, why would we throw away the fractional part?